### PR TITLE
Complete rewrite of java.util.Formatter.

### LINKED ISSUE #1619: String.format does not support thousands grouping
The following code produces different results whether run in browser vs jvm.  ScalaJS does not appear to support the thousands separator.

```
object ScalaJSExample extends js.JSApp{
  def main(): Unit = {
    println(f"${1000}%,d")    // prints 1000, not 1,000
    println(f"${1000F}%,.2f") // prints 1000.00, not 1,000.00
  }
}
```

### PR DESCRIPTION
**Pull request by sjrd**
This class was written a very long time ago, when quality standards were not as high as they are now. In this commit, we entirely rewrite it while aiming for:

* Readability
* Performance
* Fidelity in a lot of corner cases wrt. the specification on the JVM, in particular in error cases

The corresponding test suite is significantly enhanced.

In the process, this commit fixes #1619 (thousands grouping), as well as many other issues, among which:

* Formatting of astral characters with `%c`.
* Formatting of `null` with all the format specifiers.
* Handling of precision in non-numeric strings (basically `substring`).
* Capturing `IOException`s thrown by the target `Appendable`.

Despite the large number of fixes, the following items are still not addressed:

* #3202 and, more generally, various discrepancies between the formatting of floating point numbers wrt. the specification, with `%f`, `%e` and `%g`.
* Hexadecimal format for floating point numbers with `%a`.
* Formatting of date/times with the `%t.` family.
* Formatting of `BigInteger`s and `BigDecimal`s.

### CODE REVIEW

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 1]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
+          else if (flags.upperCase) "0X"
+          else "0x"
+        }
+        arg match {
+          case arg: Int =>
+            padAndSendToDest(flags, width, prefix,
+                applyUpperCase(flags, java.lang.Integer.toHexString(arg)))
+          case arg: Long =>
+            padAndSendToDest(flags, width, prefix,
+                applyUpperCase(flags, java.lang.Long.toHexString(arg)))
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Formatter.scala]: Seems like it might be worth introducing a `prefix` parameter to `formatNonNumericString`?

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 2]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
-  def ioException(): IOException = null
-  def locale(): Locale = ifNotClosed { null }
-  def out(): Appendable = ifNotClosed { dest }
+  private def applyUpperCase(flags: Flags, str: String): String =
+    if (flags.upperCase) str.toUpperCase
+    else str
 
-  override def toString(): String = out().toString()
+  /** This method ignores `flags.zeroPad` and `flags.upperCase`. */
+  private def padAndSendToDest(flags: Flags, width: Int, str: String): Unit = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Formatter.scala]: This is a bit magic, that one overload doesn't support `zeroPad` and the other one does. Could we at least `assert(!flags.zeroPad)`?

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 3]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
+    (if (flags.positivePlus) "+" else "") +
+    (if (flags.positiveSpace) " " else "") +
+    (if (flags.zeroPad) "0" else "") +
+    (if (flags.useGroupingSeps) "," else "") +
+    (if (flags.negativeParen) "(" else "") +
+    (if (flags.useLastIndex) "<" else "")
+  }
+
+  private def computerizedScientificNotation(x: Double, precision: Int,
+      forceDecimalSep: Boolean): String = {
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Formatter.scala]: call this `altFormat`? Otherwise the logic that `altFormat` means `forceDecimalSep` for efg is very implicit.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 4]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
+      computerizedScientificNotation(x, p - 1, forceDecimalSep)
+    }
+  }
+
+  private def decimalNotation(x: Double, precision: Int,
+      forceDecimalSep: Boolean): String = {
+
+    import js.JSNumberOps._
+
+    // First use JavaScript's native toExponential conversion
```
Comment (gzm0) [inline javalib/src/main/scala/java/util/Formatter.scala]: toFixed

File: javalib/src/main/scala/java/util/Formatter.scala (hunk 1 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: I considered that, but it doesn't fit. `prefix` is the thing that goes after leading space padding, but before 0 padding. If I introduce a `prefix` in `formatNonNumericString`, then I need to take 0 padding into account in that method, and that complicates things. It was cleaner in the end to only repeat `applyUpperCase` in the `x`/`X` conversion.

File: javalib/src/main/scala/java/util/Formatter.scala (hunk 3 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: I found it was more explicit to name this `forceDecimalSep` here. It is semantically related to the "notation" thing. `altFormat` is a very generic thing which applies to many different conversions, and takes on different *meanings* depending on the conversion. Once we're inside `xyzNotation`, there is a more specific meaning, with a more specific name.

I have made the `forceDecimalSep = altFormat` more explicit at call site (in `efgCommon`) to compensate.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala [hunk 5]
```diff
@@ -15,47 +15,12 @@ import org.scalajs.testsuite.utils.Platform._
...
-Double  => JDouble,
-Float   => JFloat,
-Integer => JInteger,
-Long    => JLong,
-Byte    => JByte,
-Short   => JShort,
-Boolean => JBoolean,
-String  => JString
-}
-
```
[the remark this answers is not in the document]
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala]: Done.

File: javalib/src/main/scala/java/util/Formatter.scala (hunk 2 above)
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: We actually call it when `flags.zeroPad` is true, *intentionally* disregarding it. I have made that clearer by renaming this overload to `padAndSendToDestNoZeroPad`.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 6]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
+      else "-Infinity"
+    }
+
+    padAndSendToDest(flags, width, applyUpperCase(flags, str))
+  }
+
+  private def formatNumericString(flags: Flags, width: Int, str: String): Unit = {
+
+    val TransformativeFlags =
+      PositivePlus | PositiveSpace | UseGroupingSeps | NegativeParen
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: Added a comment. It's only relevant to the fast-paths of `formatNumericString`.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 7]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
+      re.lastIndex = formatSpecifierIndex
+      val execResult = re.exec(format)
 
-          val flags = matchResult(2).get
-          def hasFlag(flag: String) = flags.indexOf(flag) >= 0
+      if (execResult == null || execResult.index != formatSpecifierIndex) {
+        // Could not parse a valid format specifier
+        val conversion =
+          if (formatSpecifierIndex == fmtLength) "%"
+          else format.substring(formatSpecifierIndex, formatSpecifierIndex + 1)
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: JVM behavior, yes. Added a comment and explicit tests.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 8]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
-            var result: String = ""
-            var i = times
-            while (i > 0) {
-              result += s
-              i -= 1
-            }
-            result
-          }
+  private def parseFlags(flags: String, conversion: Char): Flags = {
+    var bits = if (conversion <= 'Z') UpperCase else 0
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: I thought about that but I prefer it in `parseFlags`. That way, `parseFlags` contains all the logic to construct the `Flags` from various pieces of the format specifier. I don't find it strange to pass in `conversion` given that `conversion` plays a role in computing the `Flags`, at the end of the day.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 9]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
-              }
-            }
+      bits |= bit
+      i += 1
+    }
+
+    new Flags(bits)
+  }
+
+  private def parsePositiveIntSilent(capture: js.UndefOr[String],
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: It has a slight size and performance impact, yes. I have added named arguments at call site.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 10]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
+      conversion: Char, invalidFlags: Int): Unit = {
+
+    @noinline def illegalFlags(): Nothing =
+      throw new IllegalFormatFlagsException(flagsToString(flags))
+
+    if ((flags.bits & invalidFlags) != 0)
+      illegalFlags()
+  }
+
+  private def flagsToString(flags: Flags): String = {
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: Actually no. I have added a comment about that.

File: javalib/src/main/scala/java/util/Formatter.scala [hunk 11]
```diff
@@ -4,285 +4,702 @@ import scala.annotation.switch
...
-          def strRepeat(s: String, times: Int) = {
-            var result: String = ""
-            var i = times
-            while (i > 0) {
-              result += s
-              i -= 1
-            }
-            result
-          }
+  private def parseFlags(flags: String, conversion: Char): Flags = {
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalib/src/main/scala/java/util/Formatter.scala]: Same as `flagsToString`. I have added comments about that.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala [hunk 12]
```diff
@@ -67,174 +32,380 @@ class FormatterTest {
...
+
     if (!executingInJVMOnJDK6)
       expectFormatterThrows(classOf[Exception], "%#s", "hello")
   }
 
-  @Test def `should_fail_s_conversions_without_width`(): Unit = {
-    // Issue #2246
-    expectFormatterThrows(classOf[MissingFormatWidthException], "%-s", "abc")
+  @Test def sConversionWithFormattable(): Unit = {
+    assertFC("foobar", "%s", 0, -1, -1)
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala]: define `assertFC` as a local method here? It is only used in this test. Maybe even define `FormattableClass` here, also inline `"foobar`" into `assertFC`?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala [hunk 13]
```diff
@@ -67,174 +32,380 @@ class FormatterTest {
...
+
+    /* We use 1.51e100 in this test, since we seem to have a floating point
+     * imprecision at exactly 1.5e100 that yields to a rounding error towards
+     * 1e+100 instead of 2e+100.
+     */
+    assertF("2e+100", "%.0e", 1.51e100)
+
+    assertF(" 1.20e+100", "%10.2e", 1.2e100)
+
+    // Corner case of round-half-up (currently non-compliant)
```
Comment (gzm0) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala]: Should we track this as a bug?

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/util/FormatterTest.scala]: I explicitly generalized #3202 for this set of issues.
